File: ios/Runner/LocalAi/LocalAiBridge.swift
import Flutter
import Foundation
import os

#if canImport(FoundationModels)
  import FoundationModels
#endif

/// iOS side of the `dev.romlerk/local_ai` channel.
///
/// Backed by Apple's Foundation Models framework on eligible devices. On
/// everything else — an older OS, an ineligible device, Apple Intelligence
/// switched off, a model still downloading — it reports that state honestly
/// and the Dart capability router runs the deterministic parser instead.
///
/// Nothing here decides product behaviour. It reports capability, returns
/// structured drafts, and translates vendor errors into the shared code
/// taxonomy; validity rules live in Dart.
final class LocalAiBridge {
  static let channelName = "dev.romlerk/local_ai"

  /// Bumped in lockstep with `TaskParseResult.currentSchemaVersion`.
  static let schemaVersion = 1

  /// Longer than the Dart-side timeout on purpose: Dart gives up first and
  /// cancels, so this only catches a genuinely wedged session.
  private static let maxInputCharacters = 1000

  private var inFlight: [String: Task<Void, Never>] = [:]

  static func register(messenger: any FlutterBinaryMessenger) {
    let channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    let bridge = LocalAiBridge()
    channel.setMethodCallHandler { call, result in
      bridge.handle(call, result: result)
    }
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "capabilities":
      result(capabilities())

    case "parseTasks":
      guard let arguments = call.arguments as? [String: Any] else {
        result(error(.unknown, "missing arguments"))
        return
      }
      parseTasks(arguments, result: result)

    case "estimateDuration":
      // Not offered yet: the capability payload does not advertise
      // durationEstimate, so Dart never routes here.
      result(nil)

    case "cancel":
      let requestId = (call.arguments as? [String: Any])?["requestId"] as? String
      cancel(requestId)
      result(nil)

    default:
      result(FlutterMethodNotImplemented)
    }
  }

  // MARK: - Capabilities

  private func capabilities() -> [String: Any] {
    #if canImport(FoundationModels)
      if #available(iOS 26.0, *) {
        let model = SystemLanguageModel.default
        let (availability, reason): (String, String?) = {
          switch model.availability {
          case .available:
            return ("available", nil)
          case .unavailable(.deviceNotEligible):
            return ("notEligible", "DEVICE_NOT_ELIGIBLE")
          case .unavailable(.appleIntelligenceNotEnabled):
            return ("disabled", "APPLE_INTELLIGENCE_NOT_ENABLED")
          case .unavailable(.modelNotReady):
            return ("modelNotReady", "MODEL_NOT_READY")
          @unknown default:
            return ("unknown", "UNKNOWN_AVAILABILITY")
          }
        }()

        var payload: [String: Any] = [
          "provider": "appleFoundationModels",
          "availability": availability,
          "features": ["structuredText"],
          "languages": model.supportedLanguages.compactMap {
            $0.languageCode?.identifier
          },
          "constraints": [
            // Foundation Models has no foreground restriction and no quota of
            // the kind ML Kit imposes, so neither is claimed here.
            "foregroundOnly": false,
            "quotaPossible": false,
            "maxInputCharacters": Self.maxInputCharacters,
          ],
        ]
        if let reason { payload["reason"] = reason }
        return payload
      }
    #endif

    // Built without the SDK, or running on an OS that predates it.
    return [
      "provider": "appleFoundationModels",
      "availability": "notEligible",
      "features": [String](),
      "constraints": ["maxInputCharacters": Self.maxInputCharacters],
      "reason": "OS_TOO_OLD",
    ]
  }

  // MARK: - Parsing

  private func parseTasks(_ arguments: [String: Any], result: @escaping FlutterResult) {
    #if canImport(FoundationModels)
      if #available(iOS 26.0, *) {
        let text = (arguments["text"] as? String) ?? ""
        let requestId = (arguments["requestId"] as? String) ?? UUID().uuidString

        guard !text.isEmpty else {
          result(["schemaVersion": Self.schemaVersion, "tasks": [Any]()])
          return
        }
        guard text.count <= Self.maxInputCharacters else {
          result(error(.inputTooLong, "length=\(text.count)"))
          return
        }
        guard SystemLanguageModel.default.isAvailable else {
          result(error(.modelUnavailable, "model reported unavailable"))
          return
        }

        let task = Task { [weak self] in
          guard let self else { return }
          defer { self.inFlight.removeValue(forKey: requestId) }
          do {
            let payload = try await self.generate(arguments: arguments, text: text)
            guard !Task.isCancelled else {
              result(self.error(.cancelled, nil))
              return
            }
            result(payload)
          } catch is CancellationError {
            result(self.error(.cancelled, nil))
          } catch {
            result(self.translate(error))
          }
        }
        inFlight[requestId] = task
        return
      }
    #endif

    result(error(.modelUnavailable, "no on-device model on this OS"))
  }

  private func cancel(_ requestId: String?) {
    guard let requestId else { return }
    inFlight.removeValue(forKey: requestId)?.cancel()
  }

  #if canImport(FoundationModels)
    @available(iOS 26.0, *)
    private func generate(
      arguments: [String: Any],
      text: String
    ) async throws -> [String: Any] {
      let referenceNow = (arguments["referenceNow"] as? String) ?? ""
      let timezone = (arguments["timezone"] as? String) ?? TimeZone.current.identifier
      let locale = (arguments["locale"] as? String) ?? Locale.current.identifier
      let knownTags = (arguments["knownTags"] as? [String]) ?? []
      let allowMultiple = (arguments["allowMultipleTasks"] as? Bool) ?? true

      let session = LanguageModelSession(
        instructions: Self.instructions(allowMultiple: allowMultiple)
      )

      // A 3B model is poor at calendar arithmetic. Rather than ask it to work
      // out which date "Friday" is, the upcoming dates are computed here and
      // handed over as a lookup table. Measured on device, this is the
      // difference between "Friday" resolving correctly and landing three days
      // late.
      let zone = TimeZone(identifier: timezone) ?? .current
      var prompt = """
        Current time: \(referenceNow)
        Timezone: \(timezone)
        Locale: \(locale)

        Calendar (use these exact dates, do not calculate your own):
        \(Self.calendarAnchor(referenceNow: referenceNow, timezone: zone))
        """
      if !knownTags.isEmpty {
        prompt += "\n\nExisting tags: \(knownTags.joined(separator: ", "))"
      }
      prompt += "\n\nInput: \(text)"

      let response = try await session.respond(
        to: prompt,
        generating: GeneratedTaskList.self,
        options: GenerationOptions(temperature: 0.1)
      )

      return try Self.encode(
        response.content,
        allowMultiple: allowMultiple,
        timezone: TimeZone(identifier: timezone) ?? .current
      )
    }

    /// The next fortnight as an explicit weekday → date table, so resolving
    /// "Friday" is a lookup rather than arithmetic.
    private static func calendarAnchor(
      referenceNow: String,
      timezone: TimeZone
    ) -> String {
      let start =
        isoWithFraction.date(from: referenceNow)
        ?? isoPlain.date(from: referenceNow)
        ?? Date()

      var calendar = Calendar(identifier: .gregorian)
      calendar.timeZone = timezone

      let formatter = DateFormatter()
      formatter.locale = Locale(identifier: "en_US_POSIX")
      formatter.timeZone = timezone
      formatter.dateFormat = "EEEE yyyy-MM-dd"

      // Eight days covers every weekday name plus "next <weekday>" for the
      // common cases. Longer tables cost tokens, and tokens cost latency.
      return (0...8)
        .compactMap { offset -> String? in
          guard
            let day = calendar.date(byAdding: .day, value: offset, to: start)
          else { return nil }
          let label =
            switch offset {
            case 0: " (today)"
            case 1: " (tomorrow)"
            default: ""
            }
          return "- \(formatter.string(from: day))\(label)"
        }
        .joined(separator: "\n")
    }

    /// Short, task-specific, and explicit about the one rule that matters:
    /// flag rather than guess. Apple's guidance is that on-device models do
    /// better with small tasks and clear instructions than with long prompts.
    private static func instructions(allowMultiple: Bool) -> String {
      var text = """
        You convert a person's note to themselves into structured task data.

        Rules:
        - Keep the title short and actionable, in sentence case. Strip \
        scheduling words, priority words and #tags out of it.
        - For dates, copy the matching line from the supplied calendar. Never \
        compute a date yourself. Output ISO-8601 with the supplied offset.
        - If a time is vague, such as "later" or "sometime", do NOT invent one. \
        Leave the date empty and record an ambiguity explaining what is unclear.
        - Never schedule anything in the past.
        - Leave reminderAt EMPTY unless the person explicitly asked to be \
        reminded at a different moment from the due date.
        - Only use a tag the person literally typed with a # prefix. If they \
        typed none, return no tags.
        - Only set a duration the person stated. Do not estimate one.
        - Only set a priority the person signalled with words like "urgent" or \
        "!". Otherwise use none.
        """
      text +=
        allowMultiple
        ? "\n- Split genuinely separate commitments into separate tasks."
        : "\n- Produce exactly one task."
      return text
    }

    /// Converts guided-generation output into the channel payload.
    ///
    /// Timestamps are validated here: a present-but-unparseable date means the
    /// model produced something the app cannot trust, so the whole result is
    /// rejected and the deterministic parser takes over.
    @available(iOS 26.0, *)
    private static func encode(
      _ list: GeneratedTaskList,
      allowMultiple: Bool,
      timezone: TimeZone
    ) throws -> [String: Any] {
      let candidates = allowMultiple ? list.tasks : Array(list.tasks.prefix(1))

      var tasks: [[String: Any]] = []
      for task in candidates {
        let title = task.title.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !title.isEmpty else { continue }

        var encoded: [String: Any] = [
          "title": String(title.prefix(200)),
          "priority": task.priority.rawValue,
        ]

        let notes = task.notes.trimmingCharacters(in: .whitespacesAndNewlines)
        if !notes.isEmpty { encoded["notes"] = String(notes.prefix(2000)) }

        if let dueAt = try normalizeTimestamp(task.dueAt, timezone: timezone) {
          encoded["dueAt"] = dueAt
          // Default the reminder to the due time unless the model named a
          // different one; Dart drops it if it lands in the past.
          encoded["reminderAt"] =
            try normalizeTimestamp(task.reminderAt, timezone: timezone) ?? dueAt
        }

        if task.durationMinutes > 0 && task.durationMinutes <= 1440 {
          encoded["durationMinutes"] = task.durationMinutes
          encoded["durationIsEstimate"] = true
        }

        let tags = task.tags
          .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
          .filter { !$0.isEmpty }
        if !tags.isEmpty { encoded["tags"] = Array(tags.prefix(10)) }

        if task.recurrence.frequency != .none {
          var recurrence: [String: Any] = [
            "frequency": task.recurrence.frequency.rawValue,
            "interval": min(max(task.recurrence.interval, 1), 365),
          ]
          let weekdays = task.recurrence.byWeekday.filter { (1...7).contains($0) }
          if !weekdays.isEmpty { recurrence["byWeekday"] = weekdays }
          encoded["recurrence"] = recurrence
        }

        let ambiguities = task.ambiguities.compactMap { ambiguity -> [String: Any]? in
          let reason = ambiguity.reason.trimmingCharacters(in: .whitespacesAndNewlines)
          guard !reason.isEmpty else { return nil }
          var entry: [String: Any] = [
            "field": normalizeField(ambiguity.field),
            "reason": String(reason.prefix(200)),
          ]
          let span = ambiguity.sourceSpan.trimmingCharacters(in: .whitespacesAndNewlines)
          if !span.isEmpty { entry["sourceSpan"] = String(span.prefix(120)) }
          return entry
        }
        if !ambiguities.isEmpty { encoded["ambiguities"] = ambiguities }

        tasks.append(encoded)
      }

      return [
        "schemaVersion": schemaVersion,
        "provider": "appleFoundationModels",
        "tasks": Array(tasks.prefix(8)),
      ]
    }

    /// Layouts the model plausibly emits, in preference order.
    ///
    /// A small model asked for "ISO-8601 with an offset" will sometimes return
    /// a bare local time instead. That is a formatting slip, not a wrong
    /// answer, so it is repaired against the request's timezone rather than
    /// used as grounds to throw the whole result away. Anything genuinely
    /// unreadable still fails closed.
    private static let localLayouts = [
      "yyyy-MM-dd'T'HH:mm:ss",
      "yyyy-MM-dd'T'HH:mm",
      "yyyy-MM-dd HH:mm:ss",
      "yyyy-MM-dd HH:mm",
      "yyyy-MM-dd",
    ]

    private static func makeIsoFormatter(
      _ options: ISO8601DateFormatter.Options
    ) -> ISO8601DateFormatter {
      let formatter = ISO8601DateFormatter()
      formatter.formatOptions = options
      return formatter
    }

    private static let isoWithFraction = makeIsoFormatter([
      .withInternetDateTime, .withFractionalSeconds,
    ])
    private static let isoPlain = makeIsoFormatter([.withInternetDateTime])

    /// Returns nil for "no date", a normalized ISO-8601 string with an offset
    /// for anything readable, and throws only when the model claimed a date
    /// that cannot be interpreted at all.
    private static func normalizeTimestamp(
      _ raw: String,
      timezone: TimeZone
    ) throws -> String? {
      let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
      guard !trimmed.isEmpty, trimmed.lowercased() != "null" else { return nil }

      if let date = isoWithFraction.date(from: trimmed)
        ?? isoPlain.date(from: trimmed)
      {
        return output(date, timezone: timezone)
      }

      let formatter = DateFormatter()
      formatter.locale = Locale(identifier: "en_US_POSIX")
      formatter.timeZone = timezone
      for layout in localLayouts {
        formatter.dateFormat = layout
        if let date = formatter.date(from: trimmed) {
          return output(date, timezone: timezone)
        }
      }

      throw LocalAiBridgeError.invalidOutput("unparseable timestamp")
    }

    private static func output(_ date: Date, timezone: TimeZone) -> String {
      let formatter = makeIsoFormatter([.withInternetDateTime])
      formatter.timeZone = timezone
      return formatter.string(from: date)
    }

    private static func normalizeField(_ raw: String) -> String {
      switch raw.trimmingCharacters(in: .whitespacesAndNewlines) {
      case "reminderAt": return "reminderAt"
      case "recurrence": return "recurrence"
      case "startAt": return "startAt"
      case "priority": return "priority"
      case "duration": return "duration"
      case "tags": return "tags"
      case "title": return "title"
      case "notes": return "notes"
      default: return "dueAt"
      }
    }
  #endif

  // MARK: - Errors

  /// Mirrors `LocalAiErrorCode` in Dart. Every value degrades to the
  /// deterministic parser rather than surfacing as a user-visible failure.
  private enum Code: String {
    case modelUnavailable = "MODEL_UNAVAILABLE"
    case modelNotReady = "MODEL_NOT_READY"
    case busyOrQuota = "BUSY_OR_QUOTA"
    case unsupportedLanguage = "UNSUPPORTED_LANGUAGE"
    case outputInvalid = "OUTPUT_INVALID"
    case cancelled = "CANCELLED"
    case inputTooLong = "INPUT_TOO_LONG"
    case unknown = "UNKNOWN"
  }

  /// Error shape only — never user task text, since this is both logged and
  /// forwarded to the local audit table.
  /// `OSLog` rather than `Logger`, which needs iOS 14 — the app still targets
  /// iOS 13.
  private static let log = OSLog(subsystem: "dev.romlerk", category: "local_ai")

  private func error(_ code: Code, _ details: String?) -> FlutterError {
    // Marked public in the log because these are fixed identifiers written by
    // this file. No interpolated value here can carry user content.
    os_log(
      "local_ai degraded: %{public}@ %{public}@",
      log: Self.log, type: .info, code.rawValue, details ?? "-")
    return FlutterError(code: code.rawValue, message: nil, details: details)
  }

  private func translate(_ raw: Error) -> FlutterError {
    if case LocalAiBridgeError.invalidOutput(let detail) = raw {
      return error(.outputInvalid, detail)
    }

    #if canImport(FoundationModels)
      if #available(iOS 26.0, *),
        let generation = raw as? LanguageModelSession.GenerationError
      {
        switch generation {
        case .exceededContextWindowSize:
          return error(.inputTooLong, "context window")
        case .assetsUnavailable:
          return error(.modelNotReady, "assets unavailable")
        case .rateLimited, .concurrentRequests:
          return error(.busyOrQuota, "rate limited")
        case .unsupportedLanguageOrLocale:
          return error(.unsupportedLanguage, "locale unsupported")
        case .decodingFailure, .unsupportedGuide, .guardrailViolation, .refusal:
          return error(.outputInvalid, "guided generation rejected")
        @unknown default:
          return error(.unknown, "generation error")
        }
      }
    #endif

    return error(.unknown, String(describing: type(of: raw)))
  }
}

enum LocalAiBridgeError: Error {
  case invalidOutput(String)
}

File: android/app/src/main/kotlin/com/example/romlerk_mobile/LocalAiBridge.kt
package com.example.romlerk_mobile

import android.os.Build
import android.util.Log
import com.google.mlkit.genai.common.FeatureStatus
import com.google.mlkit.genai.prompt.Generation
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.text.SimpleDateFormat
import java.util.Calendar
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject

/**
 * Android side of the `dev.romlerk/local_ai` channel.
 *
 * Backed by Gemini Nano through ML Kit's GenAI Prompt API where the device
 * supports it. Everywhere else it reports that state honestly and the Dart
 * capability router runs the deterministic parser instead.
 *
 * Unlike the iOS adapter, this asks the model for JSON and parses it rather
 * than using ML Kit's typed structured output, which currently needs a KSP
 * plugin plus an alpha schema compiler. `DraftCodec` on the Dart side already
 * validates this payload as untrusted input, so malformed JSON simply becomes
 * `OUTPUT_INVALID` and the rules parser takes over — the designed behaviour,
 * not a workaround.
 */
class LocalAiBridge(messenger: BinaryMessenger) : MethodChannel.MethodCallHandler {

    private companion object {
        const val CHANNEL = "dev.romlerk/local_ai"
        const val TAG = "RomlerkLocalAi"

        /** Bumped in lockstep with `TaskParseResult.currentSchemaVersion`. */
        const val SCHEMA_VERSION = 1

        const val MAX_INPUT_CHARACTERS = 1000
        const val MAX_TITLE_LENGTH = 200
        const val MAX_TASKS = 8

        // Wire values of LocalAiErrorCode. Every one degrades to tier C.
        const val MODEL_UNAVAILABLE = "MODEL_UNAVAILABLE"
        const val MODEL_NOT_READY = "MODEL_NOT_READY"
        const val BUSY_OR_QUOTA = "BUSY_OR_QUOTA"
        const val BACKGROUND_BLOCKED = "BACKGROUND_BLOCKED"
        const val OUTPUT_INVALID = "OUTPUT_INVALID"
        const val CANCELLED = "CANCELLED"
        const val INPUT_TOO_LONG = "INPUT_TOO_LONG"
        const val UNKNOWN = "UNKNOWN"

        /**
         * ML Kit GenAI declares minSdk 26 while the app supports 24. The
         * library is force-merged in the manifest, so this guard is what
         * actually keeps an older device from ever loading those classes.
         */
        val ML_KIT_SUPPORTED = Build.VERSION.SDK_INT >= Build.VERSION_CODES.O
    }

    private val channel = MethodChannel(messenger, CHANNEL).apply {
        setMethodCallHandler(this@LocalAiBridge)
    }

    private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)
    private val inFlight = mutableMapOf<String, Job>()

    /**
     * ML Kit GenAI is foreground-only, and the BRD forbids background
     * generative work outright. MainActivity keeps this in step with the
     * activity lifecycle.
     */
    @Volatile
    var isForeground: Boolean = true

    fun dispose() {
        inFlight.values.forEach(Job::cancel)
        inFlight.clear()
        channel.setMethodCallHandler(null)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "capabilities" -> capabilities(result)
            "parseTasks" -> parseTasks(call, result)
            // Not offered: the capability payload never advertises
            // durationEstimate, so Dart does not route here.
            "estimateDuration" -> result.success(null)
            "cancel" -> {
                inFlight.remove(call.argument<String>("requestId"))?.cancel()
                result.success(null)
            }
            else -> result.notImplemented()
        }
    }

    // ---------------------------------------------------------- capabilities

    private fun capabilities(result: MethodChannel.Result) {
        scope.launch {
            val payload = mutableMapOf<String, Any?>(
                "provider" to "mlKitGenAI",
                "features" to listOf("structuredText"),
                // Gemini Nano's own language coverage is narrower than the
                // device locale list, and ML Kit does not expose it, so this
                // stays empty (unrestricted) rather than claiming support the
                // runtime has not confirmed.
                "languages" to emptyList<String>(),
                "constraints" to mapOf(
                    "foregroundOnly" to true,
                    "quotaPossible" to true,
                    "maxInputCharacters" to MAX_INPUT_CHARACTERS,
                ),
            )

            if (!ML_KIT_SUPPORTED) {
                payload["availability"] = "notEligible"
                payload["reason"] = "OS_TOO_OLD"
                result.success(payload)
                return@launch
            }

            try {
                val status = withContext(Dispatchers.IO) {
                    Generation.getClient().checkStatus()
                }
                when (status) {
                    FeatureStatus.AVAILABLE ->
                        payload["availability"] = "available"

                    // Downloadable and downloading are both "eligible but not
                    // ready" (tier B): recoverable, worth a retry, and never
                    // presented as the device being unsupported.
                    FeatureStatus.DOWNLOADABLE -> {
                        payload["availability"] = "modelNotReady"
                        payload["reason"] = "MODEL_DOWNLOADABLE"
                    }

                    FeatureStatus.DOWNLOADING -> {
                        payload["availability"] = "modelNotReady"
                        payload["reason"] = "MODEL_DOWNLOADING"
                    }

                    FeatureStatus.UNAVAILABLE -> {
                        payload["availability"] = "notEligible"
                        payload["reason"] = "DEVICE_NOT_SUPPORTED"
                    }

                    else -> {
                        payload["availability"] = "unknown"
                        payload["reason"] = "UNKNOWN_STATUS"
                    }
                }
            } catch (error: Throwable) {
                // A missing AICore, an old Play services, a throwing vendor
                // call: all just mean no enhanced understanding here.
                payload["availability"] = "notEligible"
                payload["reason"] = "STATUS_CHECK_FAILED"
                Log.i(TAG, "capability check failed: ${error.javaClass.simpleName}")
            }

            result.success(payload)
        }
    }

    // --------------------------------------------------------------- parsing

    private fun parseTasks(call: MethodCall, result: MethodChannel.Result) {
        val text = call.argument<String>("text").orEmpty()
        val requestId = call.argument<String>("requestId") ?: return run {
            fail(result, UNKNOWN, "missing requestId")
        }

        if (text.isBlank()) {
            result.success(
                mapOf("schemaVersion" to SCHEMA_VERSION, "tasks" to emptyList<Any>()),
            )
            return
        }
        if (text.length > MAX_INPUT_CHARACTERS) {
            fail(result, INPUT_TOO_LONG, "length=${text.length}")
            return
        }
        if (!isForeground) {
            fail(result, BACKGROUND_BLOCKED, "app not in foreground")
            return
        }
        if (!ML_KIT_SUPPORTED) {
            fail(result, MODEL_UNAVAILABLE, "os too old")
            return
        }

        val referenceNow = call.argument<String>("referenceNow").orEmpty()
        val timezone = call.argument<String>("timezone").orEmpty()
        val locale = call.argument<String>("locale").orEmpty()
        val knownTags = call.argument<List<String>>("knownTags").orEmpty()
        val allowMultiple = call.argument<Boolean>("allowMultipleTasks") ?: true

        val job = scope.launch {
            try {
                val client = Generation.getClient()
                val status = withContext(Dispatchers.IO) { client.checkStatus() }
                if (status != FeatureStatus.AVAILABLE) {
                    // Deliberately not triggering a download here: that is a
                    // metered, user-visible action and capture is not the
                    // moment to start one silently.
                    fail(
                        result,
                        if (status == FeatureStatus.UNAVAILABLE) MODEL_UNAVAILABLE
                        else MODEL_NOT_READY,
                        "status=$status",
                    )
                    return@launch
                }

                val prompt = buildPrompt(
                    text = text,
                    referenceNow = referenceNow,
                    timezone = timezone,
                    locale = locale,
                    knownTags = knownTags,
                    allowMultiple = allowMultiple,
                )

                val raw = withContext(Dispatchers.IO) {
                    client.generateContent(prompt)
                        .candidates
                        .firstOrNull()
                        ?.text
                        .orEmpty()
                }

                val payload = encode(raw, timezone, allowMultiple)
                result.success(payload)
            } catch (error: kotlinx.coroutines.CancellationException) {
                fail(result, CANCELLED, null)
            } catch (error: Throwable) {
                fail(result, translate(error), error.javaClass.simpleName)
            } finally {
                inFlight.remove(requestId)
            }
        }
        inFlight[requestId] = job
    }

    /**
     * Asks for JSON and nothing else.
     *
     * The upcoming dates are computed here and supplied as a lookup table:
     * a small on-device model is unreliable at calendar arithmetic, and a
     * wrong date on a reminder is the single worst failure this product has.
     */
    private fun buildPrompt(
        text: String,
        referenceNow: String,
        timezone: String,
        locale: String,
        knownTags: List<String>,
        allowMultiple: Boolean,
    ): String = buildString {
        appendLine("Convert the note into task JSON. Reply with JSON only, no prose.")
        appendLine()
        appendLine("Shape:")
        appendLine(
            """{"tasks":[{"title":"","dueAt":"","priority":"none|low|medium|high",""" +
                """"tags":[],"ambiguities":[{"field":"dueAt","reason":""}]}]}""",
        )
        appendLine()
        appendLine("Rules:")
        appendLine("- title: short, actionable, sentence case, no dates or #tags in it.")
        appendLine("- dueAt: copy a date from the calendar below. Never compute one.")
        appendLine("  Format yyyy-MM-dd'T'HH:mm:ss. Empty if no date was given.")
        appendLine("- If the time is vague (\"later\", \"sometime\"), leave dueAt empty")
        appendLine("  and add an ambiguity saying what is unclear.")
        appendLine("- Never use a past date.")
        appendLine("- tags: only words the person typed with a # prefix. Otherwise [].")
        appendLine("- priority: only if signalled with words like \"urgent\" or \"!\".")
        appendLine(
            if (allowMultiple) {
                "- Split genuinely separate commitments into separate tasks."
            } else {
                "- Produce exactly one task."
            },
        )
        appendLine()
        appendLine("Now: $referenceNow ($timezone, $locale)")
        appendLine("Calendar:")
        appendLine(calendarAnchor(referenceNow, timezone))
        if (knownTags.isNotEmpty()) {
            appendLine("Existing tags: ${knownTags.joinToString(", ")}")
        }
        appendLine()
        append("Note: $text")
    }

    private fun calendarAnchor(referenceNow: String, timezone: String): String {
        val zone = runCatching { TimeZone.getTimeZone(timezone) }
            .getOrDefault(TimeZone.getDefault())
        val start = parseTimestamp(referenceNow, zone) ?: Date()

        val labels = SimpleDateFormat("EEEE yyyy-MM-dd", Locale.US)
            .apply { this.timeZone = zone }
        val calendar = Calendar.getInstance(zone).apply { time = start }

        return (0..8).joinToString("\n") { offset ->
            val line = labels.format(calendar.time)
            val suffix = when (offset) {
                0 -> " (today)"
                1 -> " (tomorrow)"
                else -> ""
            }
            calendar.add(Calendar.DAY_OF_YEAR, 1)
            "- $line$suffix"
        }
    }

    // -------------------------------------------------------------- decoding

    /**
     * Pulls the JSON object out of whatever the model said and reshapes it to
     * the channel contract.
     *
     * Only structural repair happens here — trimming, clamping, dropping
     * empties. Anything that cannot be read at all throws, which becomes
     * `OUTPUT_INVALID` and hands over to the deterministic parser.
     */
    private fun encode(
        raw: String,
        timezone: String,
        allowMultiple: Boolean,
    ): Map<String, Any?> {
        // Models frequently wrap JSON in prose or a ``` fence, so the object is
        // extracted by brace matching rather than trusting the whole string.
        val start = raw.indexOf('{')
        val end = raw.lastIndexOf('}')
        if (start < 0 || end <= start) {
            throw IllegalArgumentException("no JSON object in response")
        }

        val root = JSONObject(raw.substring(start, end + 1))
        val rawTasks = root.optJSONArray("tasks") ?: JSONArray()
        val zone = runCatching { TimeZone.getTimeZone(timezone) }
            .getOrDefault(TimeZone.getDefault())

        val tasks = mutableListOf<Map<String, Any?>>()
        val limit = if (allowMultiple) minOf(rawTasks.length(), MAX_TASKS) else 1

        for (index in 0 until minOf(rawTasks.length(), limit)) {
            val entry = rawTasks.optJSONObject(index) ?: continue
            val title = entry.optString("title").trim()
            if (title.isEmpty()) continue

            val encoded = mutableMapOf<String, Any?>(
                "title" to title.take(MAX_TITLE_LENGTH),
                "priority" to normalizePriority(entry.optString("priority")),
            )

            normalizeTimestamp(entry.optString("dueAt"), zone)?.let { dueAt ->
                encoded["dueAt"] = dueAt
                encoded["reminderAt"] = dueAt
            }

            val tags = entry.optJSONArray("tags")
            if (tags != null) {
                val cleaned = (0 until tags.length())
                    .mapNotNull { tags.optString(it).trim().removePrefix("#").ifEmpty { null } }
                    .distinct()
                    .take(10)
                if (cleaned.isNotEmpty()) encoded["tags"] = cleaned
            }

            val ambiguities = entry.optJSONArray("ambiguities")
            if (ambiguities != null) {
                val cleaned = (0 until ambiguities.length()).mapNotNull { position ->
                    val item = ambiguities.optJSONObject(position) ?: return@mapNotNull null
                    val reason = item.optString("reason").trim()
                    if (reason.isEmpty()) return@mapNotNull null
                    mapOf(
                        "field" to normalizeField(item.optString("field")),
                        "reason" to reason.take(200),
                    )
                }
                if (cleaned.isNotEmpty()) encoded["ambiguities"] = cleaned
            }

            tasks.add(encoded)
        }

        return mapOf(
            "schemaVersion" to SCHEMA_VERSION,
            "provider" to "mlKitGenAI",
            "tasks" to tasks,
        )
    }

    private fun normalizePriority(raw: String): String = when (raw.trim().lowercase()) {
        "low" -> "low"
        "medium" -> "medium"
        "high" -> "high"
        else -> "none"
    }

    private fun normalizeField(raw: String): String = when (raw.trim()) {
        "reminderAt", "recurrence", "startAt", "priority", "duration", "tags",
        "title", "notes",
        -> raw.trim()
        else -> "dueAt"
    }

    /** Layouts the model plausibly emits, in preference order. */
    private val layouts = listOf(
        "yyyy-MM-dd'T'HH:mm:ssXXX",
        "yyyy-MM-dd'T'HH:mm:ss",
        "yyyy-MM-dd'T'HH:mm",
        "yyyy-MM-dd HH:mm:ss",
        "yyyy-MM-dd HH:mm",
        "yyyy-MM-dd",
    )

    private fun parseTimestamp(raw: String, zone: TimeZone): Date? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty()) return null
        for (layout in layouts) {
            val parsed = runCatching {
                SimpleDateFormat(layout, Locale.US)
                    .apply {
                        timeZone = zone
                        isLenient = false
                    }
                    .parse(trimmed)
            }.getOrNull()
            if (parsed != null) return parsed
        }
        return null
    }

    /**
     * Returns null for "no date" and a normalized offset timestamp otherwise.
     * A value that is present but unreadable throws, discarding the result.
     */
    private fun normalizeTimestamp(raw: String, zone: TimeZone): String? {
        val trimmed = raw.trim()
        if (trimmed.isEmpty() || trimmed.equals("null", ignoreCase = true)) return null
        val parsed = parseTimestamp(trimmed, zone)
            ?: throw IllegalArgumentException("unparseable timestamp")
        return SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ssXXX", Locale.US)
            .apply { timeZone = zone }
            .format(parsed)
    }

    // ---------------------------------------------------------------- errors

    private fun translate(error: Throwable): String {
        if (error is IllegalArgumentException || error is org.json.JSONException) {
            return OUTPUT_INVALID
        }
        val name = error.javaClass.simpleName
        return when {
            name.contains("Quota", true) || name.contains("RateLimit", true) ->
                BUSY_OR_QUOTA
            name.contains("NotFound", true) || name.contains("Unavailable", true) ->
                MODEL_UNAVAILABLE
            name.contains("Download", true) -> MODEL_NOT_READY
            else -> UNKNOWN
        }
    }

    /** Error shape only, never user task text: this reaches the audit log. */
    private fun fail(result: MethodChannel.Result, code: String, details: String?) {
        Log.i(TAG, "local_ai degraded: $code ${details ?: "-"}")
        result.error(code, null, details)
    }
}
